aptos-move/replay-benchmark/src/generator.rs

```diff
@@ -1,105 +1,166 @@
+    /// Generates a pre-block for a single block of transactions.
+    ///
+    /// Transactions are first executed on top of an actual state. Then, transactions are executed
+    /// in top of an overridden state. The outputs of two executions are compared, and the diffs
+    /// for each transaction are returned together with the inputs.
+    ///
+    /// Note: transaction execution is sequential, so that multiple inputs can be constructed in
+    /// parallel.
+    fn generate_inputs_with_diffs(
+        &self,
+        txn_block: TransactionBlock,
+    ) -> (ReadSet, Vec<TransactionDiff>) {
+        // For later comparison of outputs, find fee payers for transactions.
+        let fee_payers = txn_block
+            .transactions
+            .iter()
+            .map(|txn| {
+                txn.try_as_signed_user_txn().map(|txn| {
+                    txn.authenticator_ref()
+                        .fee_payer_address()
+                        .unwrap_or_else(|| txn.sender())
+                })
+            })
+            .collect::<Vec<_>>();
 
-    /// Partitions a sequence of transactions into blocks.
-    fn partition(&self, txns: Vec<Transaction>) -> Vec<(Version, Vec<Transaction>)> {
-        let mut begin_versions_and_blocks = Vec::with_capacity(txns.len());
+        let state_view = self.debugger.state_view_at_version(txn_block.begin_version);
+        let state_override = self.override_config.get_state_override(&state_view);
+        let workload = Workload::from(txn_block);
 
-        let mut curr_begin = self.begin_version;
-        let mut curr_block = Vec::with_capacity(txns.len());
+        // First, we execute transactions without overrides. Flush warm VM cache to ensure read-set
+        // contains modules used to warming up the VM.
+        flush_warm_vm_cache();
+        let state_view_without_override =
+            ReadSetCapturingStateView::new(&state_view, HashMap::new());
+        let onchain_outputs = execute_workload(
+            &AptosVMBlockExecutor::new(),
+            &workload,
+            &state_view_without_override,
+            1,
+        );
+        let _onchain_inputs = state_view_without_override.into_read_set();
 
-        for txn in txns {
-            if txn.is_block_start() && !curr_block.is_empty() {
-                let block_size = curr_block.len();
-                begin_versions_and_blocks.push((curr_begin, std::mem::take(&mut curr_block)));
-                curr_begin += block_size as Version;
+        // Check on-chain outputs do not modify the state we override. If so, benchmarking results
+        // may not be correct.
+        let begin = workload
+            .transaction_slice_metadata
+            .begin_version()
+            .expect("Transaction metadata must be a chunk");
+        for (idx, on_chain_output) in onchain_outputs.iter().enumerate() {
+            for (state_key, _) in on_chain_output.write_set() {
+                if state_override.contains_key(state_key) {
```

Comment:
This just writes an error but lets the execution continue, right? That's a good behavior imho for handling such case, because it could be useful to test / replay an on-chain override (based on replay) of a manually overridden setting.

Reply:
Yes, it will just log the error. This is to protect from rare cases where override was changed in the transaction range, e.g., there was a framework upgrade, or feature flag got enabled by governance proposal.

Reply:
yes, it's very good